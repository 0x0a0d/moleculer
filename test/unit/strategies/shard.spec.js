"use strict";

let ShardStrategy = require("../../../src/strategies/shard");
let ServiceBroker = require("../../../src/service-broker");
let Context = require("../../../src/context");
let { extendExpect } = require("../utils");

extendExpect(expect);

describe("Test ShardStrategy", () => {

	describe("Test constructor", () => {
		const broker = new ServiceBroker({ logger: false });
		jest.spyOn(broker.localBus, "on");

		it("test with empty opts", () => {

			let strategy = new ShardStrategy({
				opts: {
					strategyOptions: {}
				}
			}, broker);

			expect(strategy.opts).toEqual({
				shardKey: null,
				vnodes: 10,
				ringSize: null,
				cacheSize: 1000
			});

			expect(broker.localBus.on).toHaveBeenCalledTimes(1);
			expect(broker.localBus.on).toHaveBeenCalledWith("$node.**", expect.any(Function));
		});

		it("test with options", () => {
			broker.localBus.on.mockClear();

			let strategy = new ShardStrategy({
				opts: {
					strategyOptions: {
						shardKey: "#branchID",
						vnodes: 20,
						ringSize: 100,
						cacheSize: 2000
					}
				}
			}, broker);

			expect(strategy.opts).toEqual({
				shardKey: "#branchID",
				vnodes: 20,
				ringSize: 100,
				cacheSize: 2000
			});

			expect(broker.localBus.on).toHaveBeenCalledTimes(1);
			expect(broker.localBus.on).toHaveBeenCalledWith("$node.**", expect.any(Function));
		});

		it("should set needRebuild = true if '$node.**' event received", () => {
			broker.localBus.on.mockClear();

			let strategy = new ShardStrategy({
				opts: {
					strategyOptions: {}
				}
			}, broker);

			strategy.needRebuild = false;

			broker.localBus.emit("$node.connected", { id: "node-999 " });
			expect(strategy.needRebuild).toBe(true);
		});
	});

	describe("Test getKeyFromContext", () => {
		const broker = new ServiceBroker({ logger: false });
		const ctx = new Context();
		ctx.setParams({
			a: {
				b: 5
			}
		});

		ctx.meta.user = {
			name: "John"
		};

		it("should get null if shardKey is not defined", () => {

			let strategy = new ShardStrategy({
				opts: {
					strategyOptions: {}
				}
			}, broker);

			expect(strategy.getKeyFromContext(ctx)).toBeNull();
		});

		it("should get param value", () => {

			let strategy = new ShardStrategy({
				opts: {
					strategyOptions: {
						shardKey: "a.b"
					}
				}
			}, broker);

			expect(strategy.getKeyFromContext(ctx)).toBe(5);
		});

		it("should get meta value", () => {

			let strategy = new ShardStrategy({
				opts: {
					strategyOptions: {
						shardKey: "#user.name"
					}
				}
			}, broker);

			expect(strategy.getKeyFromContext(ctx)).toBe("John");
		});

		it("should call custom shardKey function", () => {
			const shardKey = jest.fn(() => "12345");
			let strategy = new ShardStrategy({
				opts: {
					strategyOptions: {
						shardKey
					}
				}
			}, broker);

			expect(strategy.getKeyFromContext(ctx)).toBe("12345");
			expect(shardKey).toHaveBeenCalledTimes(1);
			expect(shardKey).toHaveBeenCalledWith(ctx);
		});
	});

	describe("Test getHash", () => {
		const broker = new ServiceBroker({ logger: false });

		it("should calc hash", () => {
			let strategy = new ShardStrategy({
				opts: {
					strategyOptions: {}
				}
			}, broker);

			expect(strategy.getHash("John")).toBe(1631623841);
			expect(strategy.getHash("Jane")).toBe(731224371);
			expect(strategy.getHash("Adam")).toBe(2130539036);

			expect(strategy.getHash("John")).toBe(1631623841);
			expect(strategy.getHash("Jane")).toBe(731224371);
			expect(strategy.getHash("Adam")).toBe(2130539036);
		});

		it("should calc hash", () => {
			let strategy = new ShardStrategy({
				opts: {
					strategyOptions: {
						ringSize: 1000
					}
				}
			}, broker);

			expect(strategy.getHash("John")).toBe(841);
			expect(strategy.getHash("Jane")).toBe(371);
			expect(strategy.getHash("Adam")).toBe(36);

			expect(strategy.getHash("John")).toBe(841);
			expect(strategy.getHash("Jane")).toBe(371);
			expect(strategy.getHash("Adam")).toBe(36);
		});
	});

	describe("Test rebuild", () => {
		const broker = new ServiceBroker({ logger: false });
		let prevRing;
		it("should build rings", () => {
			let strategy = new ShardStrategy({
				opts: {
					strategyOptions: {}
				}
			}, broker);

			const list = [
				{ id: "node-100" },
				{ id: "node-200" },
				{ id: "node-300" },
				{ id: "node-400" },
				{ id: "node-500" },
			];

			strategy.rebuild(list);
			expect(strategy.ring).toMatchSnapshot();
			prevRing = strategy.ring;
		});

		it("should build rings with random list", () => {
			let strategy = new ShardStrategy({
				opts: {
					strategyOptions: {}
				}
			}, broker);

			const list = [
				{ id: "node-400" },
				{ id: "node-100" },
				{ id: "node-500" },
				{ id: "node-300" },
				{ id: "node-200" },
			];

			strategy.rebuild(list);
			expect(strategy.ring).toMatchSnapshot();
			expect(strategy.ring).toEqual(prevRing);
		});

		it("should build rings with custom options", () => {
			let strategy = new ShardStrategy({
				opts: {
					strategyOptions: {
						vnodes: 2,
						ringSize: 1000
					}
				}
			}, broker);

			const list = [
				{ id: "sun" },
				{ id: "mercury" },
				{ id: "venus" },
				{ id: "earth" },
				{ id: "mars" },
				{ id: "jupyter" },
				{ id: "saturn" },
				{ id: "uranus" },
				{ id: "neptunus" },
				// Poor little Pluto...
			];

			strategy.needRebuild = true;

			strategy.rebuild(list);
			expect(strategy.ring).toMatchSnapshot();
			expect(strategy.needRebuild).toBe(false);
		});

	});

	describe("Test select", () => {
		const broker = new ServiceBroker({ logger: false });
		const ctx = new Context();

		const strategy = new ShardStrategy({
			opts: {
				strategyOptions: {}
			}
		}, broker);

		strategy.getKeyFromContext = jest.fn(() => null);
		strategy.rebuild = jest.fn();
		strategy.getHash = jest.fn(() => 123456);
		strategy.getNodeIDByKey = jest.fn(() => null);

		it("should return a random item if key is null", () => {
			const list = [
				{ id: "node-100" },
				{ id: "node-200" },
				{ id: "node-300" },
				{ id: "node-400" },
				{ id: "node-500" },
			];

			expect(strategy.select(list, ctx)).toBeAnyOf(list);
			expect(strategy.getKeyFromContext).toHaveBeenCalledTimes(1);
			expect(strategy.getKeyFromContext).toHaveBeenCalledWith(ctx);
		});

		it("should return a random item if nodeID not found", () => {
			strategy.getKeyFromContext = jest.fn(() => "key");

			const list = [
				{ id: "node-100" },
				{ id: "node-200" },
				{ id: "node-300" },
				{ id: "node-400" },
				{ id: "node-500" },
			];

			expect(strategy.select(list, ctx)).toBeAnyOf(list);

			expect(strategy.getKeyFromContext).toHaveBeenCalledTimes(1);
			expect(strategy.getKeyFromContext).toHaveBeenCalledWith(ctx);

			expect(strategy.rebuild).toHaveBeenCalledTimes(1);
			expect(strategy.rebuild).toHaveBeenCalledWith(list);

			expect(strategy.getHash).toHaveBeenCalledTimes(1);
			expect(strategy.getHash).toHaveBeenCalledWith("key");

			expect(strategy.getNodeIDByKey).toHaveBeenCalledTimes(1);
			expect(strategy.getNodeIDByKey).toHaveBeenCalledWith(123456);
		});

		it("should return a specified item", () => {
			strategy.needRebuild = false;
			strategy.getKeyFromContext.mockClear();
			strategy.rebuild.mockClear();
			strategy.getHash.mockClear();
			strategy.getNodeIDByKey.mockClear();

			const list = [
				{ id: "node-100" },
				{ id: "node-200" },
				{ id: "node-300" },
				{ id: "node-400" },
				{ id: "node-500" },
			];
			strategy.getNodeIDByKey = jest.fn(() => "node-300");

			expect(strategy.select(list, ctx)).toBe(list[2]);

			expect(strategy.getKeyFromContext).toHaveBeenCalledTimes(1);
			expect(strategy.getKeyFromContext).toHaveBeenCalledWith(ctx);

			expect(strategy.rebuild).toHaveBeenCalledTimes(0);

			expect(strategy.getHash).toHaveBeenCalledTimes(1);
			expect(strategy.getHash).toHaveBeenCalledWith("key");

			expect(strategy.getNodeIDByKey).toHaveBeenCalledTimes(1);
			expect(strategy.getNodeIDByKey).toHaveBeenCalledWith(123456);
		});
	});

	describe("Test getNodeIDByKey", () => {
		const broker = new ServiceBroker({ logger: false });

		const strategy = new ShardStrategy({
			opts: {
				strategyOptions: {
					ringSize: 1000
				}
			}
		}, broker);

		const list = [
			{ id: "node-100" },
			{ id: "node-200" },
			{ id: "node-300" },
			{ id: "node-400" },
			{ id: "node-500" },
		];

		strategy.rebuild(list);

		it("should find in rings and save to cache", () => {
			expect(strategy.cache.size).toBe(0);
			expect(strategy.getNodeIDByKey(345)).toBe("node-300");
			expect(strategy.cache.size).toBe(1);
		});

		it("should give the last element", () => {
			expect(strategy.getNodeIDByKey(2000)).toBe("node-500");
		});

		it("should find in cache", () => {
			strategy.ring.length = 0;
			expect(strategy.getNodeIDByKey(345)).toBe("node-300");
			expect(strategy.cache.size).toBe(2);
			expect(strategy.getNodeIDByKey(456)).toBe(null);
		});

		it("should clear cache when rebuild called", () => {
			strategy.rebuild(list);
			expect(strategy.cache.size).toBe(0);
		});

	});

});
