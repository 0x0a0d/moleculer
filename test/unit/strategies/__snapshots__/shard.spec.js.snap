// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ShardStrategy Test rebuild should build rings 1`] = `
Array [
  Object {
    "key": 85899345,
    "nodeID": "node-100",
  },
  Object {
    "key": 171798690,
    "nodeID": "node-200",
  },
  Object {
    "key": 257698035,
    "nodeID": "node-300",
  },
  Object {
    "key": 343597380,
    "nodeID": "node-400",
  },
  Object {
    "key": 429496725,
    "nodeID": "node-500",
  },
  Object {
    "key": 515396070,
    "nodeID": "node-100",
  },
  Object {
    "key": 601295415,
    "nodeID": "node-200",
  },
  Object {
    "key": 687194760,
    "nodeID": "node-300",
  },
  Object {
    "key": 773094105,
    "nodeID": "node-400",
  },
  Object {
    "key": 858993450,
    "nodeID": "node-500",
  },
  Object {
    "key": 944892795,
    "nodeID": "node-100",
  },
  Object {
    "key": 1030792140,
    "nodeID": "node-200",
  },
  Object {
    "key": 1116691485,
    "nodeID": "node-300",
  },
  Object {
    "key": 1202590830,
    "nodeID": "node-400",
  },
  Object {
    "key": 1288490175,
    "nodeID": "node-500",
  },
  Object {
    "key": 1374389520,
    "nodeID": "node-100",
  },
  Object {
    "key": 1460288865,
    "nodeID": "node-200",
  },
  Object {
    "key": 1546188210,
    "nodeID": "node-300",
  },
  Object {
    "key": 1632087555,
    "nodeID": "node-400",
  },
  Object {
    "key": 1717986900,
    "nodeID": "node-500",
  },
  Object {
    "key": 1803886245,
    "nodeID": "node-100",
  },
  Object {
    "key": 1889785590,
    "nodeID": "node-200",
  },
  Object {
    "key": 1975684935,
    "nodeID": "node-300",
  },
  Object {
    "key": 2061584280,
    "nodeID": "node-400",
  },
  Object {
    "key": 2147483625,
    "nodeID": "node-500",
  },
  Object {
    "key": 2233382970,
    "nodeID": "node-100",
  },
  Object {
    "key": 2319282315,
    "nodeID": "node-200",
  },
  Object {
    "key": 2405181660,
    "nodeID": "node-300",
  },
  Object {
    "key": 2491081005,
    "nodeID": "node-400",
  },
  Object {
    "key": 2576980350,
    "nodeID": "node-500",
  },
  Object {
    "key": 2662879695,
    "nodeID": "node-100",
  },
  Object {
    "key": 2748779040,
    "nodeID": "node-200",
  },
  Object {
    "key": 2834678385,
    "nodeID": "node-300",
  },
  Object {
    "key": 2920577730,
    "nodeID": "node-400",
  },
  Object {
    "key": 3006477075,
    "nodeID": "node-500",
  },
  Object {
    "key": 3092376420,
    "nodeID": "node-100",
  },
  Object {
    "key": 3178275765,
    "nodeID": "node-200",
  },
  Object {
    "key": 3264175110,
    "nodeID": "node-300",
  },
  Object {
    "key": 3350074455,
    "nodeID": "node-400",
  },
  Object {
    "key": 3435973800,
    "nodeID": "node-500",
  },
  Object {
    "key": 3521873145,
    "nodeID": "node-100",
  },
  Object {
    "key": 3607772490,
    "nodeID": "node-200",
  },
  Object {
    "key": 3693671835,
    "nodeID": "node-300",
  },
  Object {
    "key": 3779571180,
    "nodeID": "node-400",
  },
  Object {
    "key": 3865470525,
    "nodeID": "node-500",
  },
  Object {
    "key": 3951369870,
    "nodeID": "node-100",
  },
  Object {
    "key": 4037269215,
    "nodeID": "node-200",
  },
  Object {
    "key": 4123168560,
    "nodeID": "node-300",
  },
  Object {
    "key": 4209067905,
    "nodeID": "node-400",
  },
  Object {
    "key": 4294967296,
    "nodeID": "node-500",
  },
]
`;

exports[`Test ShardStrategy Test rebuild should build rings with custom options 1`] = `
Array [
  Object {
    "key": 55,
    "nodeID": "earth",
  },
  Object {
    "key": 110,
    "nodeID": "jupyter",
  },
  Object {
    "key": 165,
    "nodeID": "mars",
  },
  Object {
    "key": 220,
    "nodeID": "mercury",
  },
  Object {
    "key": 275,
    "nodeID": "neptunus",
  },
  Object {
    "key": 330,
    "nodeID": "saturn",
  },
  Object {
    "key": 385,
    "nodeID": "sun",
  },
  Object {
    "key": 440,
    "nodeID": "uranus",
  },
  Object {
    "key": 495,
    "nodeID": "venus",
  },
  Object {
    "key": 550,
    "nodeID": "earth",
  },
  Object {
    "key": 605,
    "nodeID": "jupyter",
  },
  Object {
    "key": 660,
    "nodeID": "mars",
  },
  Object {
    "key": 715,
    "nodeID": "mercury",
  },
  Object {
    "key": 770,
    "nodeID": "neptunus",
  },
  Object {
    "key": 825,
    "nodeID": "saturn",
  },
  Object {
    "key": 880,
    "nodeID": "sun",
  },
  Object {
    "key": 935,
    "nodeID": "uranus",
  },
  Object {
    "key": 1000,
    "nodeID": "venus",
  },
]
`;

exports[`Test ShardStrategy Test rebuild should build rings with random list 1`] = `
Array [
  Object {
    "key": 85899345,
    "nodeID": "node-100",
  },
  Object {
    "key": 171798690,
    "nodeID": "node-200",
  },
  Object {
    "key": 257698035,
    "nodeID": "node-300",
  },
  Object {
    "key": 343597380,
    "nodeID": "node-400",
  },
  Object {
    "key": 429496725,
    "nodeID": "node-500",
  },
  Object {
    "key": 515396070,
    "nodeID": "node-100",
  },
  Object {
    "key": 601295415,
    "nodeID": "node-200",
  },
  Object {
    "key": 687194760,
    "nodeID": "node-300",
  },
  Object {
    "key": 773094105,
    "nodeID": "node-400",
  },
  Object {
    "key": 858993450,
    "nodeID": "node-500",
  },
  Object {
    "key": 944892795,
    "nodeID": "node-100",
  },
  Object {
    "key": 1030792140,
    "nodeID": "node-200",
  },
  Object {
    "key": 1116691485,
    "nodeID": "node-300",
  },
  Object {
    "key": 1202590830,
    "nodeID": "node-400",
  },
  Object {
    "key": 1288490175,
    "nodeID": "node-500",
  },
  Object {
    "key": 1374389520,
    "nodeID": "node-100",
  },
  Object {
    "key": 1460288865,
    "nodeID": "node-200",
  },
  Object {
    "key": 1546188210,
    "nodeID": "node-300",
  },
  Object {
    "key": 1632087555,
    "nodeID": "node-400",
  },
  Object {
    "key": 1717986900,
    "nodeID": "node-500",
  },
  Object {
    "key": 1803886245,
    "nodeID": "node-100",
  },
  Object {
    "key": 1889785590,
    "nodeID": "node-200",
  },
  Object {
    "key": 1975684935,
    "nodeID": "node-300",
  },
  Object {
    "key": 2061584280,
    "nodeID": "node-400",
  },
  Object {
    "key": 2147483625,
    "nodeID": "node-500",
  },
  Object {
    "key": 2233382970,
    "nodeID": "node-100",
  },
  Object {
    "key": 2319282315,
    "nodeID": "node-200",
  },
  Object {
    "key": 2405181660,
    "nodeID": "node-300",
  },
  Object {
    "key": 2491081005,
    "nodeID": "node-400",
  },
  Object {
    "key": 2576980350,
    "nodeID": "node-500",
  },
  Object {
    "key": 2662879695,
    "nodeID": "node-100",
  },
  Object {
    "key": 2748779040,
    "nodeID": "node-200",
  },
  Object {
    "key": 2834678385,
    "nodeID": "node-300",
  },
  Object {
    "key": 2920577730,
    "nodeID": "node-400",
  },
  Object {
    "key": 3006477075,
    "nodeID": "node-500",
  },
  Object {
    "key": 3092376420,
    "nodeID": "node-100",
  },
  Object {
    "key": 3178275765,
    "nodeID": "node-200",
  },
  Object {
    "key": 3264175110,
    "nodeID": "node-300",
  },
  Object {
    "key": 3350074455,
    "nodeID": "node-400",
  },
  Object {
    "key": 3435973800,
    "nodeID": "node-500",
  },
  Object {
    "key": 3521873145,
    "nodeID": "node-100",
  },
  Object {
    "key": 3607772490,
    "nodeID": "node-200",
  },
  Object {
    "key": 3693671835,
    "nodeID": "node-300",
  },
  Object {
    "key": 3779571180,
    "nodeID": "node-400",
  },
  Object {
    "key": 3865470525,
    "nodeID": "node-500",
  },
  Object {
    "key": 3951369870,
    "nodeID": "node-100",
  },
  Object {
    "key": 4037269215,
    "nodeID": "node-200",
  },
  Object {
    "key": 4123168560,
    "nodeID": "node-300",
  },
  Object {
    "key": 4209067905,
    "nodeID": "node-400",
  },
  Object {
    "key": 4294967296,
    "nodeID": "node-500",
  },
]
`;
